mantid/wfm/wfm_stitching.md

```diff
@@ -0,0 +1,169 @@
+        if e >= frame_boundaries[n].lower and e < frame_boundaries[n].upper:
+            e += frame_shifts[n]
+            break
+```
+Instead, a large number (commonly of the order of 5000) of high-resolution equally spaced bins spanning the entire range of 'arrival time at detector' are created. Each one of these bins is given a TOF shift to be applied, depending on which frame number it corresponds to. Finally, the events can be efficiently processed with fast index finding:
+```Py
+# Set up the bins
+Nbins = 5000
+bins = np.linspace(tmin, tmax, Nbins + 1)
+frame_number = np.zeros(Nbins, dtype=np.int32)
+bin_width = bins[1] - bins[0]
+for b in range(Nbins):
+    x = bins[b] + 0.5 * bin_width
+    for n in len(frame_boundaries):
+        if x >= frame_boundaries[n].lower and x < frame_boundaries[n].upper:
+            frame_number[b] = n
+            break
+
+# Perform correction to TOF
+for e in raw_events:
+    iframe = int((e - tmin) / bin_width)
+    e += frame_shifts[frame_number[iframe]]
+```
+This assumes that the number of events to be processed is much higher than `Nbins`. This can provide 5x to 10x speedups for the conversion to TOF. Some questions remain as to whether such an optimization is applicable to the case where different corrections are applied to different detector pixels if a very large number of pixels is present, as having 5000 different bins for 1M pixels does not fit in memory for common computers.
+
+**Links**
+
+- The software is available at: https://github.com/nvaytet/wfmess
+- Contact: Neil Vaytet (ESS/DMSC)
+
+## 4. In-reduction stitching vs post-processing stitching
+
+The early versions of the stitching procedure (including Method 1) depended on the Mantid software to perform the stitching, and consisted of several function calls that had to be made in a reduction workflow in order to stitch the data before further manipulation/reduction could be carried out. This was reported to be very time-consuming by several users, as they had to re-run a potentially lengthy stitching step, which was always performing the same calculations.
+
+In principle, stitching is only required to be run once, as very little tweaking of the stitching parameters should be necessary. In addition, some users were carrying out a study that was comparing WFM and single-pulse data using the same sample set-up, and were having to maintain two different (but very similar) reduction scripts, depending on whether they were reducing WFM or single-pulse data.
+
+We thus propose that WFM stitching could be implemented as a post-processing step in the data acquisition. The raw data could be read in, stitched using the parameters from the beamline (the Nexus geometry is contained inside the data file), and a new post-processed file would then be written to disk (and added to SciCat). All the parameters that were used for the stitching would be included in the file metadata. This would effectively alleviate all the issues listed above.
```

Comment:
Open question - could we consider event compression? i.e. weighting events within some pulse time tolerance to reduce file size? Or is that out of scope?

In any case I feel it is important to have BOTH the un-corrected file and the WFM corrected file as you suggest.

Reply:
I guess if file size does become an issue we could consider event comrpession. Maybe it's worth just mentioning that we have thought about this.

Reply:
I think in the first pass we wouldn't do this, but yes do mention